### PR TITLE
Normalisation is slow

### PR DESCRIPTION
**Pull request by TonioF**
Solves #392 .

I have also added that a variable's encoding is updated when the dimension order is changed during the normalisation

Checklist:

* [x] Add unit tests and/or doctests in docstrings
~~* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions~~
~~* [ ] New/modified features documented in `docs/source/*`~~
* [x] Changes documented in `docs/CHANGES.md`
* [ ] AppVeyor and Travis CI passes
* [ ] Test coverage remains or increases (target 100%)
* [ ] Associated issues closed after merge

### CODE REVIEW

**Comment by TonioF** [comment]
Tests already exist.
Second and third bullet do not apply.

File: CHANGES.md [hunk 1]
```diff
@@ -7,6 +7,7 @@
   with observe_dask_progress('Writing dataset', 100):
       dataset.to_zarr(store)  
   ```
+* Normalisation may be slow and consume much memory. (#392)
```
Comment (forman) [inline CHANGES.md:10]: That's just the symptom and not the solution. Please be clear about what has changed in xcube, so users can undestand what's new in the upcoming release.

File: xcube/core/normalize.py [hunk 2]
```diff
@@ -201,12 +201,8 @@ def _normalize_lon_360(ds: xr.Dataset) -> xr.Dataset:
...
     for var_name in var_names:
         var = ds[var_name]
-        if len(var.dims) >= 1 and var.dims[-1] == 'lon':
-            values = np.copy(var.values)
-            temp = np.copy(values[..., : lon_size_05])
-            values[..., : lon_size_05] = values[..., lon_size_05:]
-            values[..., lon_size_05:] = temp
-            new_var = xr.DataArray(values, dims=var.dims, attrs=var.attrs)
+        if 'lon' in var.dims:
+            new_var = var.roll(lon=lon_size_05, roll_coords=False)
```
Comment (forman) [inline xcube/core/normalize.py:205]: Great! But why is `roll_coords=False`? Shouldn't they be rolled too?

File: xcube/core/normalize.py [hunk 3]
```diff
@@ -726,6 +722,8 @@ def _normalize_dim_order(ds: xr.Dataset) -> xr.Dataset:
                 ds = ds.copy()
                 copy_created = True
             ds[var_name] = var.transpose(*dim_names)
+            if var.encoding:
```
Comment (forman) [inline xcube/core/normalize.py:725]: Why that condition? Anything can be in `var.encoding` so it will not be empty. On the other hand, the dictionary `var.encoding` always exists.

File: xcube/core/normalize.py [hunk 4]
```diff
@@ -726,6 +722,8 @@ def _normalize_dim_order(ds: xr.Dataset) -> xr.Dataset:
...
+                ds[var_name].encoding['chunks'] = ds[var_name].data.chunksize
```
Comment (forman) [inline xcube/core/normalize.py:726]: `ds[var_name].data.chunksize` requires that `ds[var_name].data` has an attribute `chunksize` but that cannot be assumed. It is safer to use the xarray API, but only if it is chunked:
```suggestion
            if 'chunks' in var.encoding and ds[var_name].chunks is not None:
                ds[var_name].encoding['chunks'] = ds[var_name].chunks
```

Comment (forman) [reply xcube/core/normalize.py:726]: Even this might lead to problems because dask chunking is more general than Zarr chunking. So there are dask chunks that cann not be used in Zarr. Zarr requires all but the last chunk to be same size, while dask allows for any chunk sizes greater zero.

**Review by forman — CHANGES_REQUESTED** [review]
Good fix! But there are issues still...

**Comment by forman** [comment]
> Tests already exist.
> Second and third check box do not apply.

Please use strikethrough to indicate what does not apply. If tests exists, place check.

File: xcube/core/normalize.py (hunk 2 above)
Comment (TonioF) [reply xcube/core/normalize.py:205]: No, they shouldn't. We just took the first half of the lon coordinates and put them after the second half. The purpose of the roll command is that the variables are aligned with the correct longitude values again. If we would roll the coordinates, too, the relationship between the data and the dimension would not change and stay incorrect (see test test_fix_360_lon in test_normalize . It would fail if you'd set roll_coords to True).

File: xcube/core/normalize.py (hunk 4 above)
Comment (TonioF) [reply xcube/core/normalize.py:726]: I'd like to use 

if 'chunksize' in ds[var_name].data: 

then. The reasong behind this change is that ds[var_name].encoding['chunks'] and ds[var_name].data.chunksize should have the same values, which is not correct anymore after the dimension order has changed.
What do you suggest?

Comment (forman) [reply xcube/core/normalize.py:726]: > if 'chunksize' in ds[var_name].data:

That is definitely wrong because you make the assumption that `data` implements `__contains__()`, which you can't.

File: CHANGES.md [hunk 5]
```diff
@@ -7,6 +7,7 @@
   with observe_dask_progress('Writing dataset', 100):
       dataset.to_zarr(store)  
   ```
+* Avoid accessing variable data during normalisation. (#392)
```
Comment (forman) [inline CHANGES.md:10]: Please tell the user what you mean by "normalisation" and why the change was made.

File: xcube/core/normalize.py (hunk 4 above)
Comment (forman) [reply xcube/core/normalize.py:726]: > which is not correct anymore after the dimension order has changed.

I don't get that. In this case the order of `ds[var_name].chunks` changed too. So why is there a problem?

File: xcube/core/normalize.py (hunk 3 above)
Comment (forman) [reply xcube/core/normalize.py:725]: Condition is still wrong.

**Review by forman — CHANGES_REQUESTED** [review]

File: xcube/core/normalize.py (hunk 3 above)
Comment (TonioF) [reply xcube/core/normalize.py:725]: We are still discussing what to put there instead.

File: xcube/core/normalize.py (hunk 4 above)
Comment (TonioF) [reply xcube/core/normalize.py:726]: It is required during the determination of the zarr chunks that any given chunking in the encoding corresponds to the chunking of the data sets. See here: https://github.com/pydata/xarray/blob/5735e163bea43ec9bc3c2e640fbf25a1d4a9d0c0/xarray/backends/zarr.py#L159

**Review by forman — APPROVED** [review]
Ok!
